pkg/server/data/engine/pkg.go

```diff
@@ -0,0 +1,47 @@
+// Package engine provides the core database engine
+//
+// This is a *headless* database engine that can be used to store and retrieve
+// records. It is headless because it does not communicate with the database
+// directly. Instead, it generates SQL statements that are to be executed.
+//
+// The engine is designed to be used in standalone mode or as a part of a
+// peer-to-peer network.
+//
+// # Introduction
+// The engine is designed to store and retrieve records in a table.
+// A table is a collection of records.
+// A record is a collection of fields.
+// A record has a unique id.
+// A field is a named value.
+// A value is a string, a number, or a boolean.
+//
+// The engine is designed to be used in standalone mode or as a part of a
+// peer-to-peer network.
+//
+// For each table, the engine internally creates two distinct tables:
+// * <table>
+// * <table>_history
+//
+// The <table> table is used to store the reconciled state of the records.
+// The <table>_history table is used to store the history of the records.
+//
+// When a new version of a record is created, the engine will append this
+// version inside the <table>_history table. The <table> table will be
+// updated to reflect the new version.
+//
+// When multiple versions of a record are in a conflict, the engine will
+// choose the version with the highest revision number. This is the version
+// of the record that will be reflected in the <table> table.
+//
+// The revision of a record is a string in the format <num>-<hash>.
+// The <num> is an incrementing number.
+// The <hash> is a hash of the record.
+//
+// Two records might have the same number but different hashes.
```

Comment:
if the hash is of the original record, how can two revisions of the same record have a different hash?

Reply:
or, if the hash is of the new version of the record, I don't understand yet how the revisions are linked to their original record

Reply:
let's say you and I checkout Record1 with hash `1-aaa`. We go offline. We both update the record. 

I created a new revision with hash `2-nnn` and you `2-mmm`. Both records started at `1-aaa`, and they are both at the "2nd" revision, but they have different hashes because their content is different.

Reply:
If we would for some reason update the same record `1-aaa` with the same contents on both of our sides, then we would compute the same hash, so that would be the same revision. (both you and i would get like `2-ggg` and `2-ggg`)

Reply:
like
```
# we checkout this record
{"id":"some_id", "revision":"1-aaa", "attributes: {"foo":"bar"}}
# we go offline

# you update your record to have {"foo":"baz"}, so your store contains
{"id":"some_id", "revision":"1-aaa", "attributes: {"foo":"bar"}}
{"id":"some_id", "revision":"2-bbb", "attributes: {"foo":"baz"}}

# i update my record to have {"foo":"snip"} so my store contains
{"id":"some_id", "revision":"1-aaa", "attributes: {"foo":"bar"}}
{"id":"some_id", "revision":"2-ccc", "attributes: {"foo":"snip"}}

# we go online
# we reconcile with the server, so the server contains
{"id":"some_id", "revision":"1-aaa", "attributes: {"foo":"bar"}}
{"id":"some_id", "revision":"2-bbb", "attributes: {"foo":"baz"}}
{"id":"some_id", "revision":"2-ccc", "attributes: {"foo":"snip"}}

```

Reply:
thanks for the example. the concept with the alphabetical order was quite clear, I just didn't realize that the revision is a property on the record. I thought the 2-mmm was supposed to somehow lead to the record itself, so the hash changing confused me. but if the revision is a property on the record, the hash is not relevant for the link.